### DISCUSSION TITLE
Setting next player and phase at the same time

### DISCUSSION THREAD

**Discussion by JoeLudwig** [boardgameio/boardgame.io#948] [Q&A] [answered]
My game has an end-of-round phase that applies to only a subset of players. That means I need to set a new phase and the next player at the same time. I tried to do that with:
```
            ctx.events?.endTurn?.( { next: "1" } );
            ctx.events?.setPhase?.( "calamity" );
```

It appears that "1" gets set as the current player, then that player's turn ends when setPhase ends the phase.

I also tried setting the current player with endTurn inside of onBegin in the new phase. That didn't have any effect that I could see.

The workaround I discovered was this:
```
function setPhaseAndPlayer( ctx: Ctx, phase: string, playerId: string )
{
	let i = ctx.playOrder.indexOf( playerId );
	let iMinusOne = ( i + ctx.playOrder.length - 1 ) % ctx.playOrder.length;
	ctx.events?.endTurn?.( { next: ctx.playOrder[ iMinusOne ] } );
	ctx.events?.setPhase?.( phase );
}
```

That's more than a little bit of a hack and will break if any of the above behavior changes.

Much better would be for endPhase/setPhase to take an optional player ID and sets that player as current when the phase starts.

**Comment by delucis** [comment, marked as answer]
Currently the best way to do this is using the phase’s `turn.order.first` method. If the heuristic of who starts the end-of-round phase is predictable, you can have all the logic there and be guaranteed that entering the phase always results in the correct starting player. If it’s essential to set the player from the same place as the `setPhase` event, you can store the starting player temporarily in `G` and then read that in your `first` method. (As a side note, if only some players are in this end of round phase, you may also wish to use `turn.order.playOrder` to only include those. Then your actual move logic can be as simple as `endTurn` with no need for `next` or other logic.)

Docs: https://boardgame.io/documentation/#/turn-order?id=creating-a-custom-turn-order
